File: SourceCppProject/sources/Parsers/HashGeneratorTest.cpp
#include <gtest/gtest.h>

#include "HashGenerator.hpp"


TEST(HashGeneratorTest, CalculateHashForString)
{
    const std::string test_string {"A-B-C-D-E-F-G-H"};

    const std::size_t expected_hash = std::hash<std::string>{}(test_string);
    const std::size_t actual_hash = HashGenerator::generateLineHash(test_string);

    EXPECT_EQ(actual_hash, expected_hash);
}

TEST(HashGeneratorTest, CompareHashForEqualStrings)
{
    const std::string test_string_1 {"A-B-C-D-E-F-G-H"};
    const std::string test_string_2 {"A-B-C-D-E-F-G-H"};

    const std::size_t hash_1 = HashGenerator::generateLineHash(test_string_1);
    const std::size_t hash_2 = HashGenerator::generateLineHash(test_string_2);

    EXPECT_EQ(hash_1, hash_2);
}

TEST(HashGeneratorTest, CompareHashForNotEqualStrings)
{
    const std::string test_string_1 {"A-B-C-D-E-F-G-H"};
    const std::string test_string_2 {"a-b-c-d-e-f-g-h"};

    const std::size_t hash_1 = HashGenerator::generateLineHash(test_string_1);
    const std::size_t hash_2 = HashGenerator::generateLineHash(test_string_2);

    EXPECT_NE(hash_1, hash_2);
}

TEST(HashGeneratorTest, CompareHashForCustomEqualStrings_TestCase_1)
{
    std::hash<std::string> hash {};
    EXPECT_EQ(HashGenerator::generateLineHash("0123456789"), hash("0123456789"));
}

TEST(HashGeneratorTest, CompareHashForCustomEqualStrings_TestCase_2)
{
    std::hash<std::string> hash {};
    EXPECT_EQ(HashGenerator::generateLineHash("1234567890"), hash("1234567890"));
}

TEST(HashGeneratorTest, CompareHashForCustomEqualStrings_TestCase_3)
{
    std::hash<std::string> hash {};
    EXPECT_EQ(HashGenerator::generateLineHash("2345678901"), hash("2345678901"));
}

TEST(HashGeneratorTest, CompareHashForCustomEqualStrings_TestCase_4)
{
    std::hash<std::string> hash {};
    EXPECT_EQ(HashGenerator::generateLineHash("3456789012"), hash("3456789012"));
}

TEST(HashGeneratorTest, CompareHashForCustomEqualStrings_TestCase_5)
{
    std::hash<std::string> hash {};
    EXPECT_EQ(HashGenerator::generateLineHash("4567890123"), hash("4567890123"));
}

TEST(HashGeneratorTest, CompareHashForCustomEqualStrings_TestCase_6)
{
    std::hash<std::string> hash {};
    EXPECT_EQ(HashGenerator::generateLineHash("5678901234"), hash("5678901234"));
}

TEST(HashGeneratorTest, CompareHashForCustomEqualStrings_TestCase_7)
{
    std::hash<std::string> hash {};
    EXPECT_EQ(HashGenerator::generateLineHash("6789012345"), hash("6789012345"));
}

TEST(HashGeneratorTest, CompareHashForCustomEqualStrings_TestCase_8)
{
    std::hash<std::string> hash {};
    EXPECT_EQ(HashGenerator::generateLineHash("7890123456"), hash("7890123456"));
}

TEST(HashGeneratorTest, CompareHashForCustomEqualStrings_TestCase_9)
{
    std::hash<std::string> hash {};
    EXPECT_EQ(HashGenerator::generateLineHash("8901234567"), hash("8901234567"));
}

TEST(HashGeneratorTest, CompareHashForCustomEqualStrings_TestCase_10)
{
    std::hash<std::string> hash {};
    EXPECT_EQ(HashGenerator::generateLineHash("9012345678"), hash("9012345678"));
}


TEST(HashGeneratorTest, compareHashForCustomNotEqualStrings_TestCase_1)
{
    std::hash<std::string> hash {};
    EXPECT_NE(HashGenerator::generateLineHash("012345678"), hash("123456789"));
}

TEST(HashGeneratorTest, compareHashForCustomNotEqualStrings_TestCase_2)
{
    std::hash<std::string> hash {};
    EXPECT_NE(HashGenerator::generateLineHash("123456789"), hash("234567890"));
}

TEST(HashGeneratorTest, compareHashForCustomNotEqualStrings_TestCase_3)
{
    std::hash<std::string> hash {};
    EXPECT_NE(HashGenerator::generateLineHash("234567890"), hash("345678901"));
}

TEST(HashGeneratorTest, compareHashForCustomNotEqualStrings_TestCase_4)
{
    std::hash<std::string> hash {};
    EXPECT_NE(HashGenerator::generateLineHash("345678901"), hash("456789012"));
}

TEST(HashGeneratorTest, compareHashForCustomNotEqualStrings_TestCase_5)
{
    std::hash<std::string> hash {};
    EXPECT_NE(HashGenerator::generateLineHash("456789012"), hash("567890123"));
}

TEST(HashGeneratorTest, compareHashForCustomNotEqualStrings_TestCase_6)
{
    std::hash<std::string> hash {};
    EXPECT_NE(HashGenerator::generateLineHash("567890123"), hash("678901234"));
}

TEST(HashGeneratorTest, compareHashForCustomNotEqualStrings_TestCase_7)
{
    std::hash<std::string> hash {};
    EXPECT_NE(HashGenerator::generateLineHash("678901234"), hash("789012345"));
}

TEST(HashGeneratorTest, compareHashForCustomNotEqualStrings_TestCase_8)
{
    std::hash<std::string> hash {};
    EXPECT_NE(HashGenerator::generateLineHash("789012345"), hash("890123456"));
}

TEST(HashGeneratorTest, compareHashForCustomNotEqualStrings_TestCase_9)
{
    std::hash<std::string> hash {};
    EXPECT_NE(HashGenerator::generateLineHash("890123456"), hash("901234567"));
}

TEST(HashGeneratorTest, compareHashForCustomNotEqualStrings_TestCase_10)
{
    std::hash<std::string> hash {};
    EXPECT_NE(HashGenerator::generateLineHash("901234567"), hash("012345678"));
}

File: TextLinesComparing.Testing/HashGeneratorTest.cs
using TextLinesComparing.Library;

namespace TextLinesComparing.Testing;


public class HashGeneratorTest
{
    [Test]
    public void CalculateHashForString()
    {
        const string test_string = "A-B-C-D-E-F-G-H";
    
        int expected_hash = test_string.GetHashCode();
        int actual_hash = HashGenerator.GenerateLineHash(test_string);
    
        Assert.That(actual_hash, Is.EqualTo(expected_hash));
    }

    [Test]
    public void CompareHashForEqualStrings()
    {
        const string test_string_1 = "A-B-C-D-E-F-G-H";
        const string test_string_2 = "A-B-C-D-E-F-G-H";

        int hash_1 = HashGenerator.GenerateLineHash(test_string_1);
        int hash_2 = HashGenerator.GenerateLineHash(test_string_2);

        Assert.That(actual: hash_1, expression: Is.EqualTo(hash_2));
    }

    [Test]
    public void CompareHashForNotEqualStrings()
    {
        const string test_string_1 = "A-B-C-D-E-F-G-H";
        const string test_string_2 = "a-b-c-d-e-f-g-h";
    
        int hash_1 = HashGenerator.GenerateLineHash(test_string_1);
        int hash_2 = HashGenerator.GenerateLineHash(test_string_2);
    
        Assert.That(actual: hash_1, expression: Is.Not.EqualTo(hash_2));
    }


    [Test]
    public void CompareHashForCustomEqualStrings()
    {
        Assert.Multiple(() =>
        {
            Assert.That(
            actual: HashGenerator.GenerateLineHash("0123456789"),
            expression: Is.EqualTo("0123456789".GetHashCode()));

            Assert.That(
                actual: HashGenerator.GenerateLineHash("1234567890"),
                expression: Is.EqualTo("1234567890".GetHashCode()));

            Assert.That(
                actual: HashGenerator.GenerateLineHash("2345678901"),
                expression: Is.EqualTo("2345678901".GetHashCode()));

            Assert.That(
                actual: HashGenerator.GenerateLineHash("3456789012"),
                expression: Is.EqualTo("3456789012".GetHashCode()));

            Assert.That(
                actual: HashGenerator.GenerateLineHash("4567890123"),
                expression: Is.EqualTo("4567890123".GetHashCode()));

            Assert.That(
                actual: HashGenerator.GenerateLineHash("5678901234"),
                expression: Is.EqualTo("5678901234".GetHashCode()));

            Assert.That(
                actual: HashGenerator.GenerateLineHash("6789012345"),
                expression: Is.EqualTo("6789012345".GetHashCode()));

            Assert.That(
                actual: HashGenerator.GenerateLineHash("7890123456"),
                expression: Is.EqualTo("7890123456".GetHashCode()));

            Assert.That(
                actual: HashGenerator.GenerateLineHash("8901234567"),
                expression: Is.EqualTo("8901234567".GetHashCode()));

            Assert.That(
                actual: HashGenerator.GenerateLineHash("9012345678"),
                expression: Is.EqualTo("9012345678".GetHashCode()));
        });
    }

    [Test]
    public void CompareHashForCustomNotEqualStrings()
    {
        Assert.Multiple(() =>
        {
            Assert.That(
                actual: HashGenerator.GenerateLineHash("012345678"),
                expression: Is.Not.EqualTo("123456789".GetHashCode()));

            Assert.That(
                actual: HashGenerator.GenerateLineHash("123456789"),
                expression: Is.Not.EqualTo("234567890".GetHashCode()));
            Assert.That(
                actual: HashGenerator.GenerateLineHash("234567890"),
                expression: Is.Not.EqualTo("345678901".GetHashCode()));
            Assert.That(
                actual: HashGenerator.GenerateLineHash("345678901"),
                expression: Is.Not.EqualTo("456789012".GetHashCode()));
            Assert.That(
                actual: HashGenerator.GenerateLineHash("456789012"),
                expression: Is.Not.EqualTo("567890123".GetHashCode()));
            Assert.That(
                actual: HashGenerator.GenerateLineHash("567890123"),
                expression: Is.Not.EqualTo("678901234".GetHashCode()));
            Assert.That(
                actual: HashGenerator.GenerateLineHash("678901234"),
                expression: Is.Not.EqualTo("789012345".GetHashCode()));
            Assert.That(
                actual: HashGenerator.GenerateLineHash("789012345"),
                expression: Is.Not.EqualTo("890123456".GetHashCode()));
            Assert.That(
                actual: HashGenerator.GenerateLineHash("890123456"),
                expression: Is.Not.EqualTo("901234567".GetHashCode()));
            Assert.That(
                actual: HashGenerator.GenerateLineHash("901234567"),
                expression: Is.Not.EqualTo("012345678".GetHashCode()));
        });
    }
}
